vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,407 @@
+    -------
+    True if the given contour has a contour inside of it, False otherwise
+    """
+    if hierarchy[0, contour_index, 2] <= 0:
+        return False
+    return True
+
+
+def test_min_area_box(
+    contour: contour_type, min_box_ratio: float = 0.9, max_box_ratio: float = 1.1
+) -> bool:
+    """
+    will create a box around the given contour that has the smallest possible area
+    (not necessarily upright) and check if the box's aspect ratio is within the given range
+
+    parameters
+    ----------
+    contour : contour_type
+        The individual contour to be evaluated (as returned from cv2.findContours)
+    min_box_ratio : float
+        The lowest acceptable ratio of min area box length to width must be less than the max
+    max_box_ratio : float
+        The highest ratio of min area box length to width acceptable, more than min
+
+    returns
+    -------
+    returns true if the aspect ratio of the min area box is inbetween the min and max
+    """
+    min_area_box: min_area_box_type = cv2.boxPoints(cv2.minAreaRect(contour))
+    aspect_ratio: float = (cv2.norm(min_area_box[0] - min_area_box[1])) / (
+        cv2.norm(min_area_box[1] - min_area_box[2])
+    )
+
+    return min_box_ratio < aspect_ratio < max_box_ratio
+
+
+def test_bounding_box(
+    contour: contour_type, dims: tuple[int, int], test_area_ratio: float = 10.0
+) -> bool:
+    """
+    calculates the area of a upright bounding box around the given contour
+    and compares it to the rectangle (image) of the given dimensions
+
+    parameters
+    ----------
+    contour : contour_type
+        The individual contour to be evaluated (as returned from cv2.findContours)
+    dims : tuple[int, int]
+        The dimensions of the image the contour is from
+    test_area_ratio : float
+        The minimum acceptable ratio of image area to contour bounding box area
+
+    returns
+    -------
+    returns true if the bounding box area is less than the image area by a factor of
+    test_area_ratio or more
+    """
+    bounding_box: bound_box_type = cv2.boundingRect(contour)
+
+    box_area: float = abs(bounding_box[2] - bounding_box[0]) * abs(
+        bounding_box[3] - bounding_box[1]
+    )
+    img_area: float = dims[0] * dims[1]
+
+    return box_area * test_area_ratio <= img_area
+
+
+def test_spikiness(contour: contour_type) -> bool:
+    """
+    Checks whether the contour has some points that are much farther away from the center of mass
+    than the rest of the points
+    basically real shapes will not have random points that are really far away from the rest
+    but a weird patch of grass or something without well defined edges will have crazy points
+    all over the place
+
+    parameters
+    ----------
+    contour : contour_type
+        The individual contour to be evaluated (as returned from cv2.findContours)
+
+    returns
+    -------
+    True unless there is a statistical outlier point that is uniquely far away (or close ig)
```

Comment:
How do you define outlier? Need to be a bit more descriptive/add comments where needed.